### ISSUE TITLE
Not all song lyrics are romanized (KR & CHI)

### ISSUE THREAD

**Issue by crackheadakira** [surfbryce/beautiful-lyrics#152]
The only song where I've seen this bug has been Dreamer by NCT 127 so I don't know if it's just specific to that song or not. More instances of this bug show up later on in the song.
![dreamer_nct_127](https://github.com/surfbryce/beautiful-lyrics/assets/33316699/4e8b2800-05d1-4ee6-a993-0a57c4fdbcc1)

**Comment by crackheadakira** [comment]
Actually seeing it pop-up many times now, for example in 100°C by GIRIBOY one line out of four isn't romanized, and in LOSER=LOVER by Tomorrow X Together there are many examples of 1 character not being romanized in the lyircs.
![LOSER=LOVER](https://github.com/surfbryce/beautiful-lyrics/assets/33316699/70da70c3-5441-48e0-bbd5-7762b32b2231)
![100°C](https://github.com/surfbryce/beautiful-lyrics/assets/33316699/b5e4e015-48d9-456c-a431-dd735635399a)

**Comment by crackheadakira** [comment]
I made a local "copy" of the code and tried running it and realized it was failing at the RegExp part in the `GenerateKoreanRomanization` function, so I began to search around about that which led me to this [article.](https://dev.to/dvddpl/why-is-my-regex-working-intermittently-4f4g)

So by modifying the code to what the article said it passed those checks perfectly.
![solution](https://github.com/surfbryce/beautiful-lyrics/assets/33316699/bc6e44fa-a784-465c-8876-c1710adebab4)

Gonna make a fork of this and see if it breaks anything, if not I'll make a PR.

**Comment by surfbryce** [comment]
@crackheadakira You would have to go to the dedicated Repository for these packages on my profile and locate them in the necessary location.

Also good find! You don't have to make the PR I'll go ahead and make the change since it doesn't really need to be global anyways and is just a habit.

Thanks!

**Comment by crackheadakira** [comment]
Now there is another issue, these Chinese lyrics are not being correctly identified by Franc and that results in the romanization being done with Kuroshiro instead of Pinyin. (Song is i can't by Julia Wu)
![icant_juliawu](https://github.com/surfbryce/beautiful-lyrics/assets/33316699/315233c2-bdac-415f-8391-a9c50523585f)

**Comment by crackheadakira** [comment]
My idea for a small solution is to first check if the text is Korean with the `KoreanTextTest` regex, then checking if it's Japanese by using the regex for hiragana and katakana `/([ぁ-んァ-ン])/` then last but not least checking Chinese with the [CJK](https://en.wikipedia.org/wiki/CJK_Unified_Ideographs) regex `/([\u4E00-\u9FFF])/`.
And because you've already checked if the lyrics are Japanese or Korean, there would be a small chance of them being misidentified as Chinese.

**Comment by surfbryce** [comment]
> It seems that Franc can only say 1 language that exists in the text not multiple.

@crackheadakira Franc can identify multiple languages with franc-all. So I'll have to see what languages are reported for that song and determine what a good minimum threshold would be for detecting Chinese using what we are already given.

As for your small solution we should avoid - for the most part - writing RegEx to detect the language. I myself am not a language expert and thus rely on more experienced people to do that heavy lifting for me to benefit from. Unless you are an expert in this field (for either of the three languages) I also don't recommend relying on a RegEx. Especially since it can introduce blind-spots (which I've experienced myself first hand with right-to-left language detection).

I think Kuroshiros built-in method is good enough and it's not wrong in this instance since both Chinese and Japanese do share symbols (which is why it's even able to do anything in the first place).

As for Mandarin, does your subset include modern and traditional?

So for now I want to rely on established packages built by people who understand these languages on a far greater level than I am capable of possessing. I'll take a look into seeing what franc-all returns and see if Chinese even appears. If it does I'll determine if a threshold is a good drop-in for this.

**Comment by crackheadakira** [comment]
Let me preface this by saying that the best solution would be seeing if the [i18n API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/i18n/detectLanguage), or something similar to it, is accessible through the app. In the meanwhile this is a very temporary solution I came up with that probably has a 1% failure rate, but a test suite with both franc and the RegEx should be made just in case.

> I think Kuroshiros built-in method is good enough and it's not wrong in this instance since both Chinese and Japanese do share symbols (which is why it's even able to do anything in the first place).

Whilst they do share symbols they pronounce them differently, that was the main reason I started digging into a better form of detecting the language.
Kuroshiro:
![jpnResult](https://github.com/surfbryce/beautiful-lyrics/assets/33316699/582e42c7-d2cf-4a93-b1a4-dcde75443129)
Pinyin:
![cmnResult](https://github.com/surfbryce/beautiful-lyrics/assets/33316699/1e973b6f-8ae9-474c-b51f-88e54006b746)

And I am neither a language expert or a RegEx expert so I had to rely on already written RegEx, for the hiragana/katakana I got it from [this gist,](https://gist.github.com/terrancesnyder/1345094) and Chinese from [here.](https://gist.github.com/pyliaorachel/9cbc3eeb38910f429d7bf49c0d77d07d#chinese-characters)
I did try out using the `francAll()` method beforehand but the expected language never showed up when it used 2 or more languages.

`francAll()`, the blank spots are when the language never showed up in the results:
![francAll](https://github.com/surfbryce/beautiful-lyrics/assets/33316699/9415ccae-b773-4f85-8572-210d5321792f)
The RegEx:
![regex_test](https://github.com/surfbryce/beautiful-lyrics/assets/33316699/ba66c2c5-b6f3-4c33-9205-9ae2ebd764c1)

**Comment by surfbryce** [comment]
@crackheadakira Alright well I'll move the implementation over to the RegEx you shared in this case - if it improves the detection for the time being.

I went around looking for things that can replace franc and I think I found a [very good candidate](https://www.npmjs.com/package/eld#languages)

The only things holding me back right now are these:
1. I haven't tested it, but given that franc is on their benchmark and the % they gave looks about right for franc I believe their findings
2. It lacks the Uighur, Yiddish, and Mende languages which are all right-to-left languages. I don't actually know if there any songs on Spotify that utilize these languages at the moment. I highly doubt it anyways and there's a good chance they'll be interpreted as their counterparts: Uighur as Arabic, Yiddish as Hebrew, and Mende as something (probably Arabic as well).

**Comment by surfbryce** [comment]
Alright @crackheadakira I actually melded your solution with the existing Chinese detection solution and did so with the rest of the Romanizations as well.

Korean never had a problem since its script is completely distinct from that of Japanese and Chinese. However, Japanese would sometimes be marked as Chinese (even when Chinese was the last step after Japanese and Korean).

So to cover the last remaining spots I check if the dominant language in the song is what we're aiming to Romanize. If it is then we can just carry on with romanizing the provided text since we know that's what the song is primarily.

So basically it goes like this in the Romanization if-statement:
1. Is the primary-language the language we're trying to Romanize?
2. Does the provided-text yield a positive RegEx test

This so far has proven 100% effective for every single song I've played from Japanese, Korean, and Chinese. So I'm going to assume we've got a great solution for this.

I'll list you as a contributor on the next release which should be soon. I still have some backend changes to the lyrics-server to make to improve lyric censorship (lyric variants with PG lyrics instead of Explicit) and also lyrics for BRAND new songs that get fetched when they don't exist but the next day are added and the lyrics-server doesn't recognize that.
